draft-irtf-cfrg-vdaf.md

```diff
@@ -3975,15 +3960,38 @@ def prep_shares_to_prep(Poplar1, agg_param, prep_shares):
 Aggregation parameters are valid for a given input share if no aggregation
 parameter with the same level has been used with the same input share before.
 The whole preparation phase MUST NOT be run more than once for a given
-combination of input share and level.
+combination of input share and level. This function checks that levels are
+increasing between calls, and also enforces that the prefixes at each level are
+suffixes of the previous level's prefixes.
 
 ~~~
-def is_valid(agg_param, previous_agg_params):
-    (level, _) = agg_param
-    return all(
-        level != other_level
-        for (other_level, _) in previous_agg_params
-    )
+def get_ancestor(Poplar1, input, this_level, last_level):
+    # Helper function to determine the prefix of `input` at `last_level`.
+    return input >> (this_level - last_level)
+
+def is_valid(Poplar1, agg_param, previous_agg_params):
+    # Exit early if this is the first time.
+    if len(previous_agg_params) < 1:
+      return True
+
+    (level, prefixes) = agg_param
+    (last_level, last_prefixes) = previous_agg_params[-1]
+    # The empty prefix 0 is always there.
+    last_prefixes_set = set(list(last_prefixes))
+
+    # Check that level increased.
+    if level <= last_level
+      return False
+
+    # Check that prefixes are suffixes of the last level's prefixes,
+    # unless the last level was 0 (and therefore had no prefixes).
+    if last_level > 0:
```

Comment:
I think this conditional is unnecessary, or at least the explanation in the comment is incorrect, since we already handle the first aggregation separately at the top of the function. If the last level was 0, then we would expect the last prefixes to be '0' and '1' (unless the collector had some reason to skip either). Thus, the checks below should still work, though we are high up enough in the tree that the attacker has nothing to gain in its choice of prefixes.

Reply:
The issue is the following. In the first run, there are no prefixes being passed (i.e., last_prefixes_set is empty). In the second run, either 0 or 1 or both are passed in. `get_ancestor` below will turn both of these to 0 (shifting right by 1 bit), but the membership check fails because the set is empty. The previous version incorrectly always added 0 to the set, here we just skip the check since it's supposed to always pass even on the second run (any string has the empty string as prefix).

Reply:
In the second run, "0, 1, or both" would be the prefixes from the past aggregation parameter, while the current aggregation parameter would have a level of 1 and a list of prefixes that are two bits long. The current level's prefixes get passed to `get_ancestor()` so those would be mapped to 0 or 1, which is correct.

I think the existing unit tests of `is_valid()` are misleading. For starters, they show an aggregation parameter with an empty list of candidate prefixes. This doesn't make sense, as it would produce an empty output share if the VDAF were actually run. The candidate prefixes in later aggregation parameters are shorter than they need to be, see the tests `test_poplar1` and test_generate_test_vectors` for comparison.

Reply:
Oh, I see. I guess `prefixes` confused me, I thought it refers to the nodes being expanded (the parent), but it's the nodes after the expansion. I'll fix that tomorrow, and the tests as well.